src/System.Azure.Experimental/System/Azure/Authentication.cs

```diff
@@ -210,24 +174,84 @@ public static bool TryWriteCosmosDbAuthorizationHeader(Span<byte> output, Sha256
             return true;
         }
 
-        public unsafe static byte[] ComputeKeyBytes(string key)
-        {
-            const int bufferLength = 128;
+        static readonly byte[] s_type = Encoding.UTF8.GetBytes("type=");
+        static readonly byte[] s_ver = Encoding.UTF8.GetBytes("&ver=");
+        static readonly byte[] s_sig = Encoding.UTF8.GetBytes("&sig=");
 
-            byte* pBuffer = stackalloc byte[bufferLength];
+        static readonly byte[] s_get = Encoding.UTF8.GetBytes("get\n");
+        static readonly byte[] s_post = Encoding.UTF8.GetBytes("post\n");
+        static readonly byte[] s_delete = Encoding.UTF8.GetBytes("delete\n");
+
+        const int AuthenticationHeaderBufferSize = 256;
+    }
+
+    public static class StorageAccessSignature
+    {
+        public static bool TryWrite(Span<byte> output, Sha256 hash, string verb, string canonicalizedResource, DateTime utc, out int bytesWritten)
+        {
             int written, consumed;
-            var buffer = new Span<byte>(pBuffer, bufferLength);
-            if (Utf16.ToUtf8(key.AsReadOnlySpan().AsBytes(), buffer, out consumed, out written) != TransformationStatus.Done)
+            bytesWritten = 0;
+
+            if (verb.Equals("GET", StringComparison.Ordinal))
```

Comment:
nit: can you handle the hard coded literals "GET" and its length 3 another way (I'm assuming you'll need to handle GET / PUT in several places)